wordpress-mobile/WordPress-Android
# Fixes not all post search results showing up sometimes

Fixes #11080. This PR utilizes the `PostSummary` that's introduced in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1479 to section to post list. The problem with the previous implementation is that, during `PostListItemDataSource.getItemIdentifiers` the actual models might not be in the DB, so when we rely on `PostModelTable` we end up filtering out the search results. Instead, what we should do is to keep a separate table in the DB to keep the post statuses, so we can always look them up. Please see the FluxC PR for more details.

While working on this, I made some changes to `getGroupedItemIdentifiers` to simplify it. I also ended up re-writing how the unit tests work. In my opinion, these tests are setup in a way that makes it very hard to understand and maintain them. Specifically, I feel that the reliance on verifying specific ids without them being extracted to constants (which was really not a great option here) is almost never a good option. Furthermore, in this instance specifically, ids can be either `LocalId` or `RemoteId` and relying on a specific one can be very error prone depending on how the mocking is setup.

I believe we could use a parametrized tests for all this, because essentially we are always trying to verify the same thing. Given a set of identifiers and their status, we want to check whether we add the correct section headers to the list and whether the items are put under their sections correctly. I didn't go as far as to make this a parametrized test, but I've refactored the code in a way that is pretty much how it works now, a single test runner with multiple different scenarios.

_Note: I didn't make this parametrized right now, because all tests are under `PostListItemDataSourceTest` name which indicates more than just sectioning tests. I didn't want to change that too since I made a lot more changes in this PR than I initially intended to._

**To test:**

It's a little hard to describe exact testing steps for #11080, so I am going to explain what needs to be tested instead of a simple set of instructions. Please let me know if you need help with testing.

* Start with a fresh installation so there are no posts in the DB
* Visit the posts page for a REST API powered site (even though we have `PostSummary` support for self-hosted sites, there is no search functionality, so it's not possible to test the sectioning)
* Do not scroll at all
* Go to search and type a query that you know will give you results that are not in the DB yet. You can use Stetho to check which posts you have or use a very big blog and search something old or search something in Trash
* Verify that you get the results you'd get if you searched on Calypso and posts are in the correct section. If you briefly see a post loading, that means it's working because that animation only plays if the post is not in the DB.

Having said all that, if you use post search and verify that sections are working correctly, that'd be enough because #11080 can't happen anymore with these changes.

**Here are my tests:**
* Before the fix: https://cloudup.com/cSht5x-LKkd
* After the fix: https://cloudup.com/cgILLs6j88Y
Notice how many more results I get after the fix.

**PR submission checklist:**

cc @chipsnyder
